File: WordChallenge.py
# Written by Vlad Sinitsa
# In this solution I used just a basic dictionary with a slower function to find the concatenated words but
# I multi-threaded the function to create a semi quick solution

import collections
import threading
import time

totalwords = 0
words = collections.OrderedDict()
longestword = ""
secondlongestword = ""

timestart = int(round(time.time() * 1000))

fin = open("wordsforproblem.txt","r")

# Read all the words from the file and map them into a ordered dictionary by {word: word length}
readword = fin.readline().strip()
while readword:
    words[readword] = len(readword)
    readword = fin.readline().strip()

# Sort the dictionary by word length from longest to shortest
words = collections.OrderedDict(sorted(words.iteritems(), key=lambda x: x[1], reverse=True))

# This is my thread that works through a portion of the words
class CountThread(threading.Thread):
    def __init__(self, wordstocheck):
        threading.Thread.__init__(self)
        self.wordstocheck = wordstocheck

    def run(self):
        tcount = 0
        for i in range(0, len(self.wordstocheck)):
           #print self.wordstocheck[i]
            if findword(self.wordstocheck[i], True):
                #print "im here"
                tcount += 1
        lock.acquire()  # acquire a lock on the global counter so that we can synchronize the threads
        global totalwords
        totalwords += tcount
        lock.release()


# Recursive algorithm for checking if the word is made up of other words in the file
def findword(word, first):
    for i in range(1, len(word)+1):
        if i == len(word) - 1 and first:
            return False
        else:
            if word[0:i] in words:
                tword = word[words[word[0:i]]:]
                if len(tword) == 0:
                    return True
                if findword(tword, False):
                    return True
                continue
    return False


lock = threading.Lock()  # Lock that the threads use to synchronize
threads = []
# Create the threads and give each thread 500 words then start the thread
for i in range(0, len(words)/500):
    tempThread = CountThread(list(words.keys())[i*500:(i+1)*500])
    tempThread.start()
    threads.append(tempThread)

# This is to take care of any left over words
if (len(words) % 500) > 0:
    lentemp = len(words)/500
    threadExtra = CountThread(list(words.keys())[lentemp*500:(lentemp*500)+len(words) % 500])
    threadExtra.start()
    threads.append(threadExtra)

# Finds the first and second longest words in the main thread
count = 0
while len(longestword) == 0 or len(secondlongestword) == 0:
    word = list(words.keys())[count]
    if findword(word, True):
        if len(longestword) == 0:
            longestword = word
        else:
            secondlongestword = word
    count += 1

for t in threads:
    t.join()

print "The longest word is " + longestword + " with length: " + str(len(longestword))
print "The seconds longest word is " + secondlongestword + " with length: " + str(len(secondlongestword))
print "The total count of concatenated words is: " + str(totalwords)

timeend = int(round(time.time() * 1000))

print str((timeend-timestart)/1000) + " sec"

File: WordChallenge.java
/**
 * Written by Vlad Sinitsa
 * In this solution I used a Trie tree stucture to quickly and efficiently check if each word was made up of other words in the file
 */
import java.util.ArrayList;
import java.util.HashMap;
import java.io.*;
import java.util.List;
import java.util.Scanner;
public class WordChallenge {
    public static void main(String[] args)throws IOException{
        String longestword = null;
        String secondlongestword = null;
        Scanner fin = new Scanner(new File("wordsforproblem.txt"));
        Trie tree = new Trie();
        List<String> words = new ArrayList<>();
        while(fin.hasNextLine()){
            String word = fin.nextLine();
            if(word.length() > 0) {//Is used to parse out the blank lines in the bottom of the file
                tree.insert(word);
                words.add(word);
            }
        }
        //sort the arraylist by word length from longest to shortest
        words.sort((s1, s2)-> Math.abs(s2.length()) - Math.abs(s1.length()));

        int counter = 0;
        for(String w: words){
            if(tree.isConcat(w, true)){
                if(longestword == null){//Since we sorted the list from longest to shortest we know that the first word we encounter that is a concatenated word is the longest
                    longestword = w;    //so we set it as the longest word
                }else if(secondlongestword == null){//Same goes for second longest since the first longest is no longer null
                    secondlongestword = w;
                }
                counter++;
            }
        }
        System.out.println("The longest word is: " + longestword + " with length: " + longestword.length());
        System.out.println("The second longest word is: " + secondlongestword + " with length: " + secondlongestword.length());
        System.out.println("The total count of concatenated words is: " + counter);
    }




}

class Trie{

    //Node class for Trie structure
    class Node{
        char letter;
        HashMap<Character, Node> children;
        boolean wordEnd;

        public Node(char l){
            letter = l;
            children = new HashMap<>();
            wordEnd = false;
        }

        public Node(char l, boolean isEnd){
            letter = l;
            children = new HashMap<>();
            wordEnd = isEnd;
        }

        public HashMap<Character, Node> getChildren(){return children;}
        public void setWordEnd(boolean set){wordEnd = set;}
        public boolean getWordEnd(){return wordEnd;}
    }

    //Initialize trie and other methods
    Node root;
    public Trie(){
        root = new Node(' ');
    }

    public void insert(String word){
        Node cur = root;
        for(int i = 0; i < word.length(); i++){
            HashMap<Character, Node> child = cur.getChildren();
            char temp = word.charAt(i);

            if(child.containsKey(temp)){
                cur = child.get(temp);
            } else {
                Node tempNode = new Node(word.charAt(i));
                child.put(temp, tempNode);
                cur = tempNode;
            }
        }
        cur.setWordEnd(true);
    }

    //Recursive function for determining if a word is concatenated of other words in the file
    public boolean isConcat(String word, boolean f){
        boolean first = f; //boolean variable is used to determine if were in the first recursive loop so we dont have words matching to themselves
        Node cur = root;
        for(int i = 0; i < word.length(); i++){
            HashMap<Character, Node> child = cur.getChildren();
            if(!child.containsKey(word.charAt(i))){
                return false;
            }

            cur = child.get(word.charAt(i));
            if(i == word.length()-1 && first){//will only trigger if we get to the end of a word without finding any prefixes
                return false;
            }
            if(cur.getWordEnd()){
                String temp = word.substring(i+1);
                if(temp.length() == 0){
                    return true;
                }
                if(isConcat(temp, false)){//after finding a prefix we chop the prefix off and send the shortened word back into the function
                    return true;
                }
            }
        }//If we get to the end of the for loop we know the word we are looking at is not in the file so we return false
        return false;
    }
}
